deps/GraphBLAS/Source/GxB_Matrix_tuple_iter.c

```diff
@@ -7,37 +7,80 @@
 // Sets iterator as depleted.
 static inline void _EmptyIterator
 (
-	GxB_MatrixTupleIter *iter   // Iterator to deplete.
+	_GxB_MatrixTupleIter *iter   // Iterator to deplete.
 ) {
-	iter->nvals = 0;
-	iter->nnz_idx = 0;
+	iter->nvals = 0 ;
+	iter->nnz_idx = 0 ;
 }
 
-// Create a new iterator
-GrB_Info GxB_MatrixTupleIter_new
+GrB_Info GxB_MatrixTupleIter_init
 (
-	GxB_MatrixTupleIter **iter,     // iterator to create
+	_GxB_MatrixTupleIter *iter,      // iterator to init
 	GrB_Matrix A                    // matrix to iterate over
 ) {
-	GB_WHERE(A, "GxB_MatrixTupleIter_new (A)") ;
-	GB_RETURN_IF_NULL_OR_FAULTY(A) ;
-
-	// make sure matrix is not bitmap or full
-	GxB_set(A, GxB_SPARSITY_CONTROL, GxB_SPARSE) ;
-
+	GB_WHERE(A, "GxB_MatrixTupleIter_init (iter, A)") ;
 	GrB_Index nrows ;
 	GrB_Matrix_nrows(&nrows, A) ;
+	GrB_Matrix_nvals(&iter->nvals, A) ;
+
+	GxB_Matrix_Option_get(A, GxB_SPARSITY_CONTROL, &iter->sparsity_type) ;
+	if(iter->sparsity_type != GxB_SPARSE && iter->sparsity_type != GxB_HYPERSPARSE) {
+		GB_ERROR (GrB_INVALID_VALUE,
+			"Invalid sparsity type: %d", iter->sparsity_type) ;
+	}
```

Comment:
this validation should be performed earlier, at the caller e.g. `GxB_MatrixTupleIter_new`
if this validation doesn't pass we should allocate any memory and return an error